### PR TITLE
Filter out `H` from env while unifying in `setoid_rewrite X in H`

### LINKED ISSUE #17295: Nested existential variables break expected semantics
Consider the following script.
```coq
Require Import Setoid.

Set Printing Existential Instances.

Axiom A : Type.
Axiom i : A.
Axiom Equal : A -> A -> Prop.
Axiom P : A -> Prop.

Axiom everything_equal : forall (j : A), Equal i j.

Instance b : Morphisms.Proper (Equal ==> Basics.impl) P. Admitted.

Lemma blah (H : P i) : True.
rewrite everything_equal in H.
Show Proof.
(*
(fun H : P i =>
 ?Goal@{H:=(fun lemma : Equal i ?Goal0@{H:=H} => b i ?Goal0@{H:=H} lemma)
             (everything_equal ?Goal0@{H:=H}) H}) 
*)
Show.
(* 
  H : P ?Goal0@{H:=H}
  ============================
  True 
*)
exact I.
Unshelve. exact i.
Qed.
```
The printed proof term looks acceptable at first sight. However, the printed proof state violates the semantics that I've always ascribed to existential instances. Granted, I'm not sure that a formal description of existential variables for Coq exists. However, I've always interpreted it such that terms in an existential instance should be typable in the current context. As such, in the existential variable `?Goal0@{H:=H}` of hypothesis `H` of the printed proof state, I would expect the instance term `H` to be typable in an empty context, which it isn't.

Am I interpreting the semantics wrong, or is this a bug?

### PR DESCRIPTION
**Pull request by LasseBlaauwbroek**
Fix #17295

This is technically speaking probably a breaking change. But I'd expect the damage to be minimal.

- [x] Added / updated **test-suite**.

- [ ] Added **changelog**.

### CODE REVIEW

**Comment by LasseBlaauwbroek** [comment]
@coqbot run full ci

**Comment by LasseBlaauwbroek** [comment]
Mhm, it looks like this approach is too heavy-handed and will cause a massive amount of breakage. Any suggestions for a more subtle solution?

File: tactics/rewrite.ml [hunk 1]
```diff
@@ -1612,15 +1624,7 @@ let cl_rewrite_clause_newtac ?abs ?origsigma ~progress strat clause =
     | None -> concl
     | Some id -> EConstr.of_constr (Environ.named_type id env)
     in
-    let env = match clause with
-    | None -> env
-    | Some id ->
-      (* Only consider variables not depending on [id] *)
-      let ctx = named_context env in
-      let filter decl = not (occur_var_in_decl env sigma id decl) in
```
Comment (SkySkimmer) [inline tactics/rewrite.ml:1620]: This filter removes variables depending on `id` but not `id` itself, if you fix that (without other changes) it seems to work
~~~ocaml
      let filter decl = not (Id.equal id (NamedDecl.get_id decl)) && not (occur_var_in_decl env sigma id decl) in
~~~
(also update the comment 2 lines above)

**Comment by ppedrot** [comment]
This looks incorrect to me, even if we clear the hypotheses depending on the variable. We have no guarantee that the conclusion is still well-typed. Basically we should call the low-level code for the clear tactic instead.

File: tactics/rewrite.ml (hunk 1 above)
Comment (LasseBlaauwbroek) [reply tactics/rewrite.ml:1620]: Isn't that exactly what this PR is doing?

Comment (LasseBlaauwbroek) [reply tactics/rewrite.ml:1620]: Note that the other changes are needed to make `rewrite` work. Otherwise, only `setoid_rewrite` is fixed.

**Comment by LasseBlaauwbroek** [comment]
The bigger problem here seems to be that with this PR, `H` gets removed from all the subgoals, not only shelved goals. Although that is not necessarily wrong, it causes a lot of breakage. Many proofs seem to depend on the original `H` to still be there in generated side-conditions. And this PR removes that.

File: tactics/rewrite.ml (hunk 1 above)
Comment (SkySkimmer) [reply tactics/rewrite.ml:1620]: >Isn't that exactly what this PR is doing?

yes, I misread

**Comment by LasseBlaauwbroek** [comment]
Closed in favor of https://github.com/coq/coq/pull/17304

**Pull request closed without merging**